File: frontend/src/api/integrations.ts
import { api } from "./client";

/**
 * Connected systems (§26).
 *
 * Three things to keep in mind reading these types.
 *
 * **`state` is derived; `status` is not sent.** Whether an integration has what
 * it needs is a fact about its settings, and the stored column used to
 * contradict it — three of twelve rows said `NOT_CONFIGURED` while holding
 * every setting they required. The server computes it, so the browser has
 * nothing to reconcile.
 *
 * **`enabled` is intent and `state` is outcome.** Which makes "switched on and
 * failing" a state that exists — and the row an operator most needs, so the
 * page must not smooth the two into one word.
 *
 * **A check does not contact the provider, and says so.** `reached_provider`
 * is always `false` in this template: nothing here holds real credentials, and
 * a screen reporting "Connected to Stripe" when no packet left the process
 * would be the worst possible lie on it. The type says `false` rather than
 * `boolean` so a reader cannot miss it.
 *
 * See `services/integrations.py`.
 */

export type IntegrationState =
  | "NOT_CONFIGURED"
  | "DISCONNECTED"
  | "CONNECTED"
  | "ERROR";

export interface Integration {
  id: string;
  key: string;
  name: string;
  provider: string;
  category: string;
  description: string | null;
  /** What somebody asked for. */
  enabled: boolean;
  /** What is actually true, derived — see the module note. */
  state: IntegrationState;
  configured: boolean;
  /** Exactly which required settings are absent. */
  missing_settings: string[];
  required_settings: string[];
  last_connected_at: string | null;
  last_error: string | null;
  last_error_at: string | null;
  icon: string | null;
  docs_url: string | null;
}

export interface IntegrationDetail extends Integration {
  /** Values whose *names* say they are sensitive come back as the redaction. */
  configuration: Record<string, unknown>;
  /** Which keys were redacted, so the page can say "set, not shown". */
  redacted_settings: string[];
}

export interface IntegrationsPage {
  items: Integration[];
  total: number;
  page: number;
  page_size: number;
  pages: number;
  facets: Record<string, Array<{ value: string; count: number }>>;
  columns: string[];
}

export interface IntegrationCatalogue {
  fields: Array<{ name: string; label: string; kind: string }>;
  default_columns: string[];
  categories: Array<{ key: string; count: number }>;
  /** Counted on the *derived* state, so the chips agree with the rows. */
  states: Array<{ key: IntegrationState; count: number }>;
  total: number;
  /** Switched on and not connected — the reason to open this page. */
  needing_attention: number;
  /** The string a redacted value reads as, so the page can recognise it. */
  redacted: string;
}

/** What a configuration check established, and what it did not. */
export interface CheckResult extends Integration {
  /** Always `false` here. See the module note. */
  reached_provider: false;
  checked_at: string;
  note: string;
}

export interface IntegrationQuery {
  [key: string]: unknown;
  q?: string;
  category?: string;
  status?: string;
  page?: number;
  page_size?: number;
}

export const integrationsApi = {
  catalogue: (signal?: AbortSignal) =>
    api.get<IntegrationCatalogue>("/admin/integrations/catalogue", { signal }),
  list: (params: IntegrationQuery = {}, signal?: AbortSignal) =>
    api.get<IntegrationsPage>("/admin/integrations", { params, signal }),
  entry: (id: string, signal?: AbortSignal) =>
    api.get<IntegrationDetail>(`/admin/integrations/${id}`, { signal }),
  /**
   * Merged, not replaced. Sending a redacted value back means "leave it
   * alone" — otherwise showing a masked token once would destroy it.
   */
  configure: (id: string, configuration: Record<string, unknown>) =>
    api.put<IntegrationDetail>(`/admin/integrations/${id}`, { configuration }),
  /** Refused while required settings are missing, with those named. */
  setEnabled: (id: string, enabled: boolean) =>
    api.put<Integration>(`/admin/integrations/${id}/enabled`, { enabled }),
  /** Verifies the configuration. Does not contact the provider. */
  check: (id: string) => api.post<CheckResult>(`/admin/integrations/${id}/check`, {}),
};

File: backend/src/api/integrations.py
"""Integration endpoints (§26).

One permission, `integrations.manage`: the configuration names hosts and secret
references, which is reconnaissance even where it is not a secret.

No create and no delete. The providers this platform knows how to talk to come
with the code — a list an operator extends by deploying, not by typing a name.
What they can do is configure, enable, and check.

`check` is worth reading the service docstring for: it verifies the
configuration and says in the payload that it did *not* contact the provider.
A button reporting "Connected to Stripe" when no packet left the process would
be the worst possible lie on this screen.
"""

from __future__ import annotations

from typing import Any

from src.core.auth import me, requires
from src.core.db import session_scope
from src.services import integrations as service


@requires("integrations.manage")
def catalogue(app=None, operation: str = "", request=None, **_: Any):
    """Categories and states with a count against each, and what needs attention."""
    with session_scope() as session:
        return service.catalogue(session, principal=me()), 200


@requires("integrations.manage")
def collection(app=None, operation: str = "", request=None, **_: Any):
    """Every integration, filtered and faceted in PostgreSQL."""
    args = request.args if request is not None else {}
    with session_scope() as session:
        return service.listing(session, args, principal=me()), 200


@requires("integrations.manage")
def item(app=None, operation: str = "", request=None, integration_id: str = "", **kwargs: Any):
    """One integration with its configuration — or a change to those settings.

    Values whose *names* say they hold something sensitive come back redacted,
    and sending the redaction back means "leave it alone": a form that had
    never shown the real token would otherwise overwrite it with bullets.
    """
    identifier = integration_id or str(kwargs.get("integration_id") or "")
    method = (request.method if request is not None else "GET").upper()
    with session_scope() as session:
        if method == "PUT":
            payload = (request.get_json(silent=True) if request is not None else None) or {}
            return service.configure(session, identifier, payload, principal=me()), 200
        return service.entry(session, identifier, principal=me()), 200


@requires("integrations.manage")
def enablement(
    app=None, operation: str = "", request=None, integration_id: str = "", **kwargs: Any
):
    """Turn an integration on or off (§26).

    Refused while it is missing settings: switching on something that cannot
    possibly work produces a failure with no cause to find, and the refusal
    names the settings instead.
    """
    identifier = integration_id or str(kwargs.get("integration_id") or "")
    payload = (request.get_json(silent=True) if request is not None else None) or {}
    with session_scope() as session:
        return service.set_enabled(session, identifier, payload, principal=me()), 200


@requires("integrations.manage")
def check(app=None, operation: str = "", request=None, integration_id: str = "", **kwargs: Any):
    """Verify the configuration, and say what was and was not tried (§26).

    Carries `reached_provider: false`, because it is a configuration check and
    not a handshake. See the service docstring for where to make it real.
    """
    identifier = integration_id or str(kwargs.get("integration_id") or "")
    with session_scope() as session:
        return service.check(session, identifier, principal=me()), 200
